### PR TITLE
Fix WMS referrer not sent

### PR DESCRIPTION
**Pull request by benoitdm-oslandia**
Fix issue #48532 

* add constructor without QgsSettings, using default
* add new const QgsHttpHeaders::KEY_REFERER
* add test for QgsOwsConnection
* QgsOwsConnection:
  * fix usage of "referer" by using QgsHttpHeaders
  * replace "qgis/" prefix by QgsSettings::Prefix::QGIS

### CODE REVIEW

**Comment by rouault** [comment]
(note: it would help reviewers if you put a more descriptive text in the pull request title)

**Comment by agiudiceandrea** [comment]
I've tested this PR (MingW64 build), but it seems to me there are still some issue.

If you add a new WMS service with a referer URL, then the referer header is correctly sent with the request, but the referer URL is no longer visible in the WMS/WMTS Connection windows.
If you add additional headers to the WMS service, then these additional headers are not sent with the request and are no longer visible in the WMS/WMTS Connection windows.

**Comment by benoitdm-oslandia** [comment]
@agiudiceandrea thanks for the tests! I have fixed some strange behaviors with the QgsSettings key format and the table widget.

I only have a Qt 5.15.2, against what version have you tested?

**Comment by agiudiceandrea** [comment]
> I only have a Qt 5.15.2, against what version have you tested?

Qt version used is 5.15.3.
With the new push, the referer URL is correctly sent and it is visible in the WMS/WMTS Connection window.
Anyway, any additional header added is not sent with the request, while now it is visible in the WMS/WMTS Connection window.

**Comment by benoitdm-oslandia** [comment]
> Anyway, any additional header added is not sent with the request, while now it is visible in the WMS/WMTS Connection window.

@agiudiceandrea indeed I have forgotten how the referer is handled in the other parts of QGis. To fix this I will have to fix all usage of QgsDataSourceUri (for example) to handle http headers and add wrappers to not breaking the API.

**Comment by agiudiceandrea** [comment]
> @agiudiceandrea indeed I have forgotten how the referer is handled in the other parts of QGis. To fix this I will have to fix all usage of QgsDataSourceUri (for example) to handle http headers and add wrappers to not break the API.

I think that until the additional headers are not sent with the requests for any provider, then the Advanced tab in the HTTP Headers section should be hidden or disabled for that provider.

**Comment by agiudiceandrea** [comment]
Thank you very much for you effort!

**Comment by agiudiceandrea** [comment]
@benoitdm-oslandia, I've tested the WMS provider with this PR, and it seems to me that all the reported issue are now fixed: the referer URL and any additional header added in the Advanced tab of the HTTP Headers section are correctly sent with the GetCapabilities request and they are visible in the WMS/WMTS Connection window.

**Comment by nyalldawson** [comment]
Is this still a Wip?

**Comment by benoitdm-oslandia** [comment]
@nyalldawson 

> Is this still a Wip?

 No! I fix the conflicts!

**Comment by benoitdm-oslandia** [comment]
@nyalldawson ready to merge

File: src/core/qgsdatasourceuri.h [hunk 1]
```diff
@@ -325,6 +326,23 @@ class CORE_EXPORT QgsDataSourceUri
      */
     QSet<QString> parameterKeys() const;
 
+#ifndef SIP_RUN
+    //! Returns http headers
+    QgsHttpHeaders httpHeaders() const { return mHttpHeaders; }
+#endif
+
+    //! Returns http headers
```
Comment (nyalldawson) [inline src/core/qgsdatasourceuri.h:334]: Add since 3.26 to these new methods

File: src/gui/qgsmanageconnectionsdialog.cpp [hunk 2]
```diff
@@ -875,9 +883,12 @@ void QgsManageConnectionsDialog::loadOWSConnections( const QDomDocument &doc, co
     settings.setValue( QString( '/' + connectionName + "/ignoreGetFeatureInfoURI" ), child.attribute( QStringLiteral( "ignoreGetFeatureInfoURI" ) ) == QLatin1String( "true" ) );
     settings.setValue( QString( '/' + connectionName + "/ignoreAxisOrientation" ), child.attribute( QStringLiteral( "ignoreAxisOrientation" ) ) == QLatin1String( "true" ) );
     settings.setValue( QString( '/' + connectionName + "/invertAxisOrientation" ), child.attribute( QStringLiteral( "invertAxisOrientation" ) ) == QLatin1String( "true" ) );
-    settings.setValue( QString( '/' + connectionName + "/referer" ), child.attribute( QStringLiteral( "referer" ) ) );
     settings.setValue( QString( '/' + connectionName + "/smoothPixmapTransform" ), child.attribute( QStringLiteral( "smoothPixmapTransform" ) ) == QLatin1String( "true" ) );
     settings.setValue( QString( '/' + connectionName + "/dpiMode" ), child.attribute( QStringLiteral( "dpiMode" ), QStringLiteral( "7" ) ).toInt() );
+
+    QgsHttpHeaders httpHeader( child );
+    httpHeader.updateSettings( settings, QString( '/' + connectionName ) );
```
Comment (nyalldawson) [inline src/gui/qgsmanageconnectionsdialog.cpp:890]: Does this handle upgrading the old format files automatically?

**Comment by nyalldawson** [comment]
@benoitdm-oslandia I can't say I fully follow all the changes here, but I can't spot any issues in the code 👍

File: src/gui/qgsmanageconnectionsdialog.cpp (hunk 2 above)
Comment (benoitdm-oslandia) [reply src/gui/qgsmanageconnectionsdialog.cpp:890]: Yes! And it still produces old data format for the referer information.

File: src/core/qgsdatasourceuri.h (hunk 1 above)
Comment (benoitdm-oslandia) [reply src/core/qgsdatasourceuri.h]: fixed

**Review by nyalldawson — APPROVED** [review]

**Pull request merged**